Fix part of #3132: add i18n keys to About and Profile pages; minor Hindi translations for the search bar

rohan-gulati opened the pull request:
This PR adds translations to the About (```/about```) page and the Profile (```/profile```) page. I've also added minor Hindi translations, that were missing from the drop down menu(s) on the Library search bar.

kevinlee12 commented:
Hi @rohan-gulati, there seems to be some backend test issues, can you take a look at them? thanks!

kevinlee12 commented:
It looks like you might have missed a spot see: https://travis-ci.org/oppia/oppia/jobs/219593000#L3820

kevinlee12 commented on assets/i18n/en.json:
```diff
@@ -1,5 +1,40 @@
+    "I18N_ABOUT_PAGE_ABOUT_TAB_FEEDBACK": "Earn feedback",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_FEEDBACK_TEXT": "to improve your exploration.",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_HEADING": "About Oppia",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_PARAGRAPH_1": "Oppia's mission is to help anyone learn anything they want in an effective and enjoyable way.",
```
In the event where a language's paragraphs don't match English's (ie. A language might be able to accomplish the same meaning in 2 paragraphs whereas English has to have 6 paragraphs or vice versa), it's better for the paragraphs to be explicitly defined (ie. insert `<p>` tags in the JSON) within a larger `18N_ABOUT_PAGE_ABOUT_TAB_PARAGRAPH`.

rohan-gulati replied:
Done

kevinlee12 commented on assets/i18n/qqq.json:
```diff
@@ -1,5 +1,40 @@
+	"I18N_ABOUT_PAGE_ABOUT_TAB_FEEDBACK": "Text shown at the bottom of the About section of Oppia's About page.",
+	"I18N_ABOUT_PAGE_ABOUT_TAB_FEEDBACK_TEXT": "Text shown at the bottom of the About section of Oppia's About page.",
+	"I18N_ABOUT_PAGE_ABOUT_TAB_HEADING": "Heading for the About section of Oppia's About page.",
+	"I18N_ABOUT_PAGE_ABOUT_TAB_PARAGRAPH_1": "Text for the About section of Oppia's About page.",
```
Ditto

rohan-gulati replied:
Done

kevinlee12 commented on core/templates/dev/head/pages/about/about.html:
```diff
@@ -26,94 +26,59 @@
-                Oppia's mission is to help anyone learn anything
-                they want in an effective and enjoyable way.
+
+              <p translate="I18N_ABOUT_PAGE_ABOUT_TAB_PARAGRAPH_1">
```
Combine these into one larger `div`

rohan-gulati replied:
Done

wxyxinyu replied:
nit: I think this paragraph tag doesn't have a closing tag? It's not necessary, but aids consistency and maintainability (and other paragraph tags on the page do have closing tags).

rohan-gulati commented on core/controllers/pages_test.py:
```diff
@@ -24,4 +24,3 @@ def test_about_page(self):
         response = self.testapp.get('/about')
         self.assertEqual(response.status_int, 200)
         self.assertEqual(response.content_type, 'text/html')
-        response.mustcontain('Credits', 'License')
```
@kevinlee12 This is the back end test that was failing. It specifically checks the About page for these strings, which I've moved to translation keys. I've removed this part of the test for now, but I'd really appreciate your take on this.

kevinlee12 replied:
Since translations is a frontend item, we might want to tweak the tests here a little. Would it work if you check that the i18n tag is being passed to the front?

kevinlee12 replied:
You missed this one :smile:

kevinlee12 commented:
@rohan-gulati, and also please also resolve the merge conflicts. thanks!

rohan-gulati commented:
@kevinlee12 I've tweaked the test to check for  i18n keys. I've also resolved the merge conflicts. PTAL.
Thanks!

kevinlee12 commented:
Hi @wxyxinyu, can you give this a second pass? thanks!

rohan-gulati commented on core/controllers/pages_test.py:
```diff
@@ -24,4 +24,6 @@ def test_about_page(self):
-        response.mustcontain('Credits', 'License')
+        response.mustcontain(
+            'I18N_ABOUT_PAGE_CREDITS_TAB_HEADING',
+            'I18N_ABOUT_PAGE_FOUNDATION_TAB_LICENSE')
```
@kevinlee12 I think this fixes the test. PTAL, thanks!

kevinlee12 commented:
/cc @seanlip

kevinlee12 commented:
Can either @seanlip or @wxyxinyu give a second pass on this? My main concern is big About text section, ideally it should be one big chunk and it'll be up the the translators to separate it; however, I have a feeling something is off but I'm not sure how to describe it. Thanks!

seanlip commented on assets/i18n/en.json:
```diff
@@ -1,4 +1,26 @@
+    "I18N_ABOUT_PAGE_ABOUT_TAB_HEADING": "About Oppia",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_SHARE": "Publish &amp; Share",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_SHARE_TEXT": "your creations with the community.",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_TEXT": "<p>Oppia's mission is to help anyone learn anything they want in an effective and enjoyable way.</p><p>By creating a set of free, high-quality, demonstrably effective lessons with the help of educators from around the world, Oppia aims to provide students with quality education &mdash; regardless of where they are or what traditional resources they have acce ...
```
Hi @Nikerabbit @siebrand @nemobis -- just wanted to check a quick thing: is this text OK for a translation entry, or is it too long? If the latter, do you have any advice on what the best thing to do would be?

Thanks!

seanlip commented:
@rohan-gulati, would you consider handling hi.json as well?

rohan-gulati commented:
@seanlip Although Hindi is my first language, I don't think my translations would do justice to the text, since it conveys a very specific message, and I am not very confident in my translation abilities.

siebrand commented:
If you could split this up in paragraphs and insert the translations in a template, that would be preferable

siebrand commented:
That would also make partial translations possible

rohan-gulati commented:
@siebrand This is what I did in my [initial commit](https://github.com/oppia/oppia/pull/3299/commits/5250969e292f8509a40b1b3ba3c54cf57106a2a0).
PTAL! Is this the correct way to do this?

seanlip commented:
@rohan-gulati re hi.json, I would say go ahead anyway, if that's OK with you? If there are issues, the translations can always be updated via translatewiki, but it's much easier to update an existing translation than to compose one from scratch. I think this is also relevant given the work we're currently doing in India.

Also, @siebrand, thanks for weighing in!

siebrand commented:
translatewiki.net isn't great for wall of text. Splitting up per paragraph makes sense for multiple reasons (platform, translators having to deal with HTML). It's my standard advice, despite the argument made in parent that some languages might need fewer paragraphs to convey the same meaning.

rohan-gulati commented:
@seanlip I'll get on right away. Would you rather I commit that to this PR, or do it with my next (which would hopefully close #3132)?
@siebrand Thanks!

seanlip commented:
I think doing it with this one is better. Thanks!

On Monday, April 10, 2017, Rohan Gulati <notifications@github.com> wrote:

> @seanlip <https://github.com/seanlip> I'll get on right away. Would you
> rather I commit that to this PR, or do it with my next (which would
> hopefully close #3132 <https://github.com/oppia/oppia/issues/3132>)?
> @siebrand <https://github.com/siebrand> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/3299#issuecomment-292892878>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeynG807G6r5EHItVNUEEw8wFMaBF8ks5rufIegaJpZM4M2mJz>
> .
>

rohan-gulati commented:
@seanlip I'll get started, thanks!
Also, have we reached a consensus on how to go about large blocks of text? Should I split them into paragraphs, or have one large block of text?

seanlip commented:
Follow @siebrand's advice. Thanks!

rohan-gulati commented:
In this commit, I have:
- Gone back to the older way of handling larger blocks of text, i.e using separate smaller paragraphs.
- Added Hindi translations for the About page
- Modified ```pages_test.py``` to check for the i18n tags containing the strings that the older test was checking ('Credits', 'License')

@seanlip PTAL, thanks!

seanlip commented on assets/i18n/en.json:
```diff
@@ -1,4 +1,41 @@
+    "I18N_ABOUT_PAGE_ABOUT_TAB_HEADING": "About Oppia",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_PARAGRAPH_1": "Oppia's mission is to help anyone learn anything they want in an effective and enjoyable way.",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_PARAGRAPH_2": "By creating a set of free, high-quality, demonstrably effective lessons with the help of educators from around the world, Oppia aims to provide students with quality education &mdash; regardless of where they are or what traditional resources they have access to.",
+    "I18N_ABOUT_PAGE_ABOUT_TAB_PARAGRAPH_3": "So far, educators have created more than 8,000 of these lessons, which we call <em>explorations</em>. And they serve almost 250,000 students worldwide.",
```
Actually, it's now 11,000 and more than 430,000.

Which brings up an interesting question. Can we find a way to be able to update these numbers in the future without sending the whole thing for re-translation?

rohan-gulati replied:
Done.
(This was an interesting challenge, but since angular-translate provides capabilities for variable replacement, I've used that to pass  the values. This way we can change it without affecting the rest of the text.)

rohan-gulati replied:
Done. 
(I've used variable replacement in angular-translate, to store the numeric values separately.)

seanlip commented on assets/i18n/en.json:
```diff
@@ -1,4 +1,41 @@
+    "I18N_ABOUT_PAGE_ABOUT_TAB_SHARE_TEXT": "your creations with the community.",
+    "I18N_ABOUT_PAGE_CREDITS_TAB_HEADING": "Credits",
+    "I18N_ABOUT_PAGE_CREDITS_TAB_TEXT": "Oppia's contributors come from all over the world &mdash; many of us are students, recent students and teachers. We'd like to thank the following contributors who've helped to build the platform. If you'd like to help, <a href=\"/contact\">here's how you can get involved</a>!",
+    "I18N_ABOUT_PAGE_CREDITS_TAB_TEXT_BOTTOM": "The Oppia development team is also grateful for the feedback, ideas, help and suggestions by Alex Kauffmann, Allison Barros, Amy Latten, Brett Barros, Crystal Kwok, Daniel Hernandez, Divya Siddarth, Ilwon Yoon, Jennifer Chen, John Cox, John Orr, Katie Berlent, Michael Wawszczak, Mike Gainer, Neil Fraser, Noah Falstein, Nupur Jain, Peter Norvig, Phil ...
```
We also update this list of names periodically.

rohan-gulati replied:
Done.

seanlip commented on assets/i18n/en.json:
```diff
@@ -1,4 +1,41 @@
+    "I18N_ABOUT_PAGE_FOUNDATION_TAB_PARAGRAPH_5": "All content in Oppia's explorations is licensed under <a href=\"https://creativecommons.org/licenses/by-sa/4.0/legalcode\" class=\"inline-links\">CC-BY-SA 4.0</a> with a waiver of the attribution requirement (specifically, Sections 3(a)(1) and 3(a)(2)). If you reuse content from this site, we encourage you to include a link to the relevant explor ...
+    "I18N_ABOUT_PAGE_FOUNDATION_TAB_PARAGRAPH_5_LICENSE_HEADING": "License",
+    "I18N_ABOUT_PAGE_FOUNDATION_TAB_PARAGRAPH_6": "The software powering Oppia is open source, and its <a href=\"https://github.com/oppia/oppia/\" class=\"inline-links\">code</a> is released under an <a href=\"https://www.apache.org/licenses/\" class=\"inline-links\">Apache 2.0</a> license.",
+    "I18N_ABOUT_PAGE_HEADING":"Oppia: education for all",
```
Nit, here and elsewhere: add single space after colon.

rohan-gulati replied:
Fixed this.

seanlip commented on assets/i18n/hi.json:
```diff
@@ -1,5 +1,42 @@
+	"I18N_ABOUT_PAGE_FOUNDATION_TAB_PARAGRAPH_4": "फाउंडेशन के निदेशक अमित ड्यूश, जेकब डेविस, और शॉन लिप हैं। फाउंडेशन के <a href=\"https://drive.google.com/folderview?id=0B9KSjiibL_WDfjlpUUpqampZSTBOTmRqM2IyaGN3R3NlVWstZlhKcjlZQ3VMTHJBQWNGTFU&usp=sharing\" class=\"inline-links\">बायलॉज़ और मिनट </a>पढ़ने के लिए उपलब्ध हैं। यदि आप फाउंडेशन से संपर्क करना चाहते हैं, तो कृपया ईमेल करें: <a href=\"mail ...
+	"I18N_ABOUT_PAGE_FOUNDATION_TAB_PARAGRAPH_4_HEADING": "निदेशक",
+	"I18N_ABOUT_PAGE_FOUNDATION_TAB_PARAGRAPH_5": "ओपिया के अन्वेषणों की सभी सामग्री <a href=\"https://creativecommons.org/licenses/by-sa/4.0/legalcode\" class=\"inline-links\">सीसी-बाय-एसए 4.0 </a>के तहत लाइसेंस की आवश्यकता के छूट (विशेष रूप से, धारा 3 (ए) (1) और 3 (ए) (2)) के अंतर्गत लाइसेंस प्राप्त है। यदि आप इस साइट से सामग्री का पुन: उपयोग करते हैं, तो हम आपको प्रासंगिक अन्वेषण पृष्ठ के लिंक को ...
+	"I18N_ABOUT_PAGE_FOUNDATION_TAB_PARAGRAPH_5_LICHEADING": "लाइसेंस",
```
The ID seems wrong. This is a simple fix, but I am concerned about why it was not caught by either tests or manual testing. Could you investigate?

rohan-gulati replied:
I've fixed this. Also, I did try to find out why this happened, but couldn't come to a conclusion.

wxyxinyu replied:
hmm, this is kind of puzzling. It'll be really good if you could look into it again -- I believe there should be a test that checks that all the keys are in the qqq file. Could you check that it's actually being run, etc?

seanlip commented on core/templates/dev/head/pages/about/about.html:
```diff
@@ -26,94 +26,59 @@
-                  <h3>Earn feedback</h3>
-                  <p style="margin-top: -10px;">
-                    to improve your exploration.
+                  <h3 translate="I18N_ABOUT_PAGE_ABOUT_TAB_FEEDBACK">Earn feedback</h3>
```
Should "Earn feedback" still be there?

rohan-gulati replied:
No, it shouldn't; fixed it.

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -1,4 +1,41 @@
 {
+	"I18N_ABOUT_PAGE_ABOUT_TAB_CREATE": "Text shown at the bottom of the About section of Oppia's About page.",
+	"I18N_ABOUT_PAGE_ABOUT_TAB_CREATE_TEXT": "Text shown at the bottom of the About section of Oppia's About page.",
```
I think you might want to provide some info to translators (who may not actually see the Oppia website) of how this message relates to the previous one. It's not clear that it is a continuation. I also wonder how well this construct would work in other languages.

rohan-gulati replied:
Done.

rohan-gulati commented:
@seanlip PTAL. Thanks, and sorry for the delay!

seanlip commented on core/templates/dev/head/pages/about/About.js:
```diff
@@ -1,13 +1,13 @@
 // Copyright 2016 The Oppia Authors. All Rights Reserved.
 //
-// Licensed under the Apache License, Version 2.0 (the "License");
+// Licensed under the Apache License, Version 2.0 (the 'License');
```
Don't change the license text.

rohan-gulati replied:
Fixed this.

seanlip commented on core/templates/dev/head/pages/about/About.js:
```diff
@@ -7,7 +7,7 @@
 //
 // Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS-IS" BASIS,
+// distributed under the License is distributed on an 'AS-IS' BASIS,
```
Here, too.

rohan-gulati replied:
Done

seanlip commented on core/templates/dev/head/pages/about/About.js:
```diff
@@ -19,6 +19,19 @@
 oppia.controller('About', [
   '$scope', 'UrlInterpolationService',
   function($scope, UrlInterpolationService) {
+    $scope.listOfNames = ' Alex Kauffmann,' + ' Allison Barros,' +
```
Should store this data in an actual list.

rohan-gulati replied:
I tried doing this initially, but angular's ```ng-repeat``` (used to iterate through the list) does not seem to work when added as part of inline HTML inside the text for the i18n keys. I'd really appreciate some help on this. 
Thanks!

seanlip replied:
Use a list (defined as var) to store the names. Declare a separate $scope.xxx = listOfNames.join(', ') (though you may need to add some logic for the & bit).

rohan-gulati replied:
Done. Thanks for the tip!

seanlip commented:
(Please note that there are merge conflicts.)

seanlip commented:
Thanks Rohan. I can't review this now -- will pass to @wxyxinyu.

rohan-gulati commented:
ping @wxyxinyu

rohan-gulati commented:
Thanks @wxyxinyu. Added the missing tag.
